### PR TITLE
Add Transactional trait for atomic DB operations

### PR DESCRIPTION
**Pull request by ChristophWurst**
RFC https://help.nextcloud.com/t/api-proposal-trait-for-transactional-operations/133907

Ref https://github.com/nextcloud/mail/pull/6030#discussion_r806939201

Documentation: https://github.com/nextcloud/documentation/pull/8048

### CODE REVIEW

**Review by nickvergessen — APPROVED** [review]
LGTM it's quite simple and might work for most places at least

File: lib/public/AppFramework/Db/Transactional.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+use OCP\DB\Exception;
+use OCP\IDBConnection;
+use Throwable;
+
+/**
+ * Helper trait for transactional operations
+ *
+ * @since 24.0.0
+ */
+trait Transactional {
```
Comment (come-nc) [inline lib/public/AppFramework/Db/Transactional.php:37]: I’ve seen all interfaces being named IStuff and traits TThing, so this should be TTransactional I think?

File: lib/public/AppFramework/Db/Transactional.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	 * - Commit if no exceptions are thrown
+	 * - Revert otherwise and rethrows the exception
+	 *
+	 * @template T
+	 * @param callable $fn
+	 * @psalm-param callable():T $fn
+	 * @param IDBConnection $db
+	 *
+	 * @return mixed
+	 * @psalm-return T
```
Comment (come-nc) [inline lib/public/AppFramework/Db/Transactional.php:51]: Please add a small description of what is returned by the method

**Review by come-nc — CHANGES_REQUESTED** [review]

**Review by come-nc — APPROVED** [review]

**Review by CarlSchwan — APPROVED** [review]
Very nice 👍

**Pull request merged**